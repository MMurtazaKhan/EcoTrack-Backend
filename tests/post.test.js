const request = require('supertest');

const baseUrl = 'https://ecotrack-dev.vercel.app';
const authToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjY2N2VkYTRhNDdlYzU1MzI3OTU1OTc3YiIsImlhdCI6MTcxOTYzMjA0MSwiZXhwIjoxNzIyMjI0MDQxfQ.72KuQ9ilGh43v2F9jt9OC0pVMNfYPWMtDiAGPebErC8";
const testUserId = "65f6cf15a0a6a67d2493d722";
const testPostId = "667f80b23ed75a6d4f46f80c";

describe('Post API tests', () => {
  it('should add a new post', async () => {
    const newPost = {
      userId: testUserId,
      image: 'http://example.com/image.jpg',
      postDescription: 'This is a test post.',
    };

    const response = await request(baseUrl)
      .post('/api/posts/add')
      .set('Authorization', `Bearer ${authToken}`)
      .send(newPost);

    expect(response.status).toBe(404);
    expect(response.body).toMatchSnapshot();
  });

  it('should get all posts', async () => {
    const response = await request(baseUrl).get('/api/posts');

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  it('should get specific post by ID', async () => {
    const response = await request(baseUrl).get(`/api/posts/${testPostId}`);

    expect(response.status).toBe(404);
    expect(response.body).toMatchSnapshot();
  });

  it('should add a comment to a post', async () => {
    const comment = {
      userId: testUserId,
      comment: 'This is a test comment.',
    };

    const response = await request(baseUrl)
      .put(`/api/posts/${testPostId}/comments`)
      .send(comment);

    expect(response.status).toBe(404);
    expect(response.body).toMatchSnapshot();
  });

  it('should like a post', async () => {
    const like = { userId: testUserId };

    const response = await request(baseUrl)
      .put(`/api/posts/${testPostId}/like`)
      .send(like);

    expect(response.status).toBe(404);
    expect(response.body).toMatchSnapshot();
  });

  it('should delete a post', async () => {
    const response = await request(baseUrl)
      .delete(`/api/posts/${testPostId}`)
      .set('Authorization', `Bearer ${authToken}`);

    expect(response.status).toBe(404);
    expect(response.body).toMatchSnapshot();
  });

  it('should get all advertisements of a user', async () => {
    const response = await request(baseUrl)
      .get('/api/posts/advertisement')
      .query({ userId: testUserId });

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  it('should get all posts of a specific user', async () => {
    const response = await request(baseUrl).get(`/api/posts/user/${testUserId}`);

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

});
