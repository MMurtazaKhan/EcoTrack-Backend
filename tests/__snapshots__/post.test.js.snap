// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post API tests should add a comment to a post 1`] = `
{
  "message": "User not found",
}
`;

exports[`Post API tests should add a new post 1`] = `
{
  "message": "User not found",
}
`;

exports[`Post API tests should delete a post 1`] = `
{
  "message": "Post not found",
}
`;

exports[`Post API tests should get all advertisements of a user 1`] = `[]`;

exports[`Post API tests should get all posts 1`] = `
[
  {
    "__v": 5,
    "_id": "666c0eedb33b5ed0562dfc6c",
    "comments": [
      {
        "_id": "666dbce635497066efb2c585",
        "comment": "Nice Ad",
        "createdAt": "2024-06-15T16:10:14.342Z",
        "user": null,
      },
      {
        "_id": "666dbd0c35497066efb2c5a5",
        "comment": "Why Ad?",
        "createdAt": "2024-06-15T16:10:52.548Z",
        "user": null,
      },
    ],
    "createdAt": "2024-06-14T09:35:41.538Z",
    "image": "https://firebasestorage.googleapis.com/v0/b/eco-track-37da4.appspot.com/o/images%2Fvoucher.jpg?alt=media&token=901431d7-7239-4342-acdc-b2243644b512",
    "isTypeAd": false,
    "likes": [
      "65f6cf15a0a6a67d2493d722",
    ],
    "shares": [],
    "updatedAt": "2024-06-18T06:49:30.823Z",
    "user": null,
  },
  {
    "__v": 2,
    "_id": "666944a2d8761d7fde2a5e5e",
    "comments": [
      {
        "_id": "666d8b56d337baa088f190a3",
        "comment": "Great initiative ",
        "createdAt": "2024-06-15T12:38:46.628Z",
        "user": null,
      },
    ],
    "createdAt": "2024-06-12T06:48:02.251Z",
    "image": "https://firebasestorage.googleapis.com/v0/b/eco-track-37da4.appspot.com/o/adb38d5b-67db-49d8-a08f-266567a60542.jpeg?alt=media&token=8aea0c44-b50f-42ce-a53f-9fe5b5cbdb4f",
    "isTypeAd": false,
    "likes": [
      "65f6cf15a0a6a67d2493d722",
    ],
    "postDescription": "Hey guys help Us to make our city green against 
",
    "shares": [],
    "updatedAt": "2024-06-15T12:38:46.629Z",
    "user": null,
  },
  {
    "__v": 4,
    "_id": "6655ab23e25bcfa530d9c27c",
    "comments": [
      {
        "_id": "665923dd60f0ce5aa4e62313",
        "comment": "Good jerry",
        "createdAt": "2024-05-31T01:11:57.899Z",
        "user": null,
      },
    ],
    "createdAt": "2024-05-28T10:00:03.329Z",
    "image": "https://firebasestorage.googleapis.com/v0/b/eco-track-37da4.appspot.com/o/24bc5260-81b3-4fb3-b754-72dd3e135149.jpeg?alt=media&token=03573cf5-847b-4426-9019-2964f7fa2fb5",
    "isTypeAd": false,
    "likes": [
      "65f6cf15a0a6a67d2493d722",
    ],
    "postDescription": "Hi",
    "shares": [],
    "updatedAt": "2024-06-08T09:55:56.800Z",
    "user": null,
  },
  {
    "__v": 6,
    "_id": "666d74e6996bbd57c3948d1d",
    "comments": [
      {
        "_id": "666dc5343ab77f2fc2d2732f",
        "comment": "Wow",
        "createdAt": "2024-06-15T16:45:40.147Z",
        "user": "65f6cf15a0a6a67d2493d722",
      },
      {
        "_id": "666dc5383ab77f2fc2d2733c",
        "comment": "Nice",
        "createdAt": "2024-06-15T16:45:44.824Z",
        "user": "65f6cf15a0a6a67d2493d722",
      },
      {
        "_id": "666dc53e3ab77f2fc2d2734b",
        "comment": "Very dangerous ",
        "createdAt": "2024-06-15T16:45:50.386Z",
        "user": "65f6cf15a0a6a67d2493d722",
      },
      {
        "_id": "666f028faf1b2f0381a166ab",
        "comment": "How can I buy this?",
        "createdAt": "2024-06-16T15:19:43.924Z",
        "user": "65f6cf15a0a6a67d2493d722",
      },
    ],
    "createdAt": "2024-06-15T11:03:02.846Z",
    "image": "https://firebasestorage.googleapis.com/v0/b/eco-track-37da4.appspot.com/o/images%2Fcrone.png?alt=media&token=ae142bd0-78d2-48cd-9378-a586cfc7a84f",
    "isTypeAd": true,
    "likes": [],
    "postDescription": "Introducing the EcoCharge Solar Power Bank: harness clean energy on the go and keep your devices charged sustainably. Join the green revolution today!",
    "shares": [],
    "updatedAt": "2024-06-16T15:38:37.582Z",
    "user": null,
  },
  {
    "__v": 33,
    "_id": "662fbc8ff9e8c5468deeca7f",
    "comments": [
      {
        "_id": "664450f4501eb2dc65217435",
        "comment": "Yes ",
        "createdAt": "2024-05-15T06:06:44.454Z",
        "user": null,
      },
      {
        "_id": "6652fb3ebe8826f705f240e9",
        "comment": "Nice",
        "createdAt": "2024-05-26T09:05:02.809Z",
        "user": null,
      },
      {
        "_id": "665322deb5e3e687884722c3",
        "comment": "Kya post hai bhai",
        "createdAt": "2024-05-26T11:54:06.751Z",
        "user": null,
      },
      {
        "_id": "6653237a45df2da31e41fbba",
        "comment": "Good",
        "createdAt": "2024-05-26T11:56:42.007Z",
        "user": null,
      },
      {
        "_id": "6653245645df2da31e41fbd6",
        "comment": "Nice",
        "createdAt": "2024-05-26T12:00:22.557Z",
        "user": null,
      },
      {
        "_id": "665325b3b5e3e687884722ea",
        "comment": "Hi",
        "createdAt": "2024-05-26T12:06:11.239Z",
        "user": null,
      },
      {
        "_id": "66532640b5e3e6878847230c",
        "comment": "Try",
        "createdAt": "2024-05-26T12:08:32.296Z",
        "user": null,
      },
      {
        "_id": "665326e245df2da31e41fc12",
        "comment": "Wheo",
        "createdAt": "2024-05-26T12:11:14.032Z",
        "user": null,
      },
      {
        "_id": "6653273db5e3e6878847233e",
        "comment": "Thoko thoko",
        "createdAt": "2024-05-26T12:12:45.668Z",
        "user": null,
      },
      {
        "_id": "665327bdb5e3e6878847236b",
        "comment": "Fuck",
        "createdAt": "2024-05-26T12:14:53.975Z",
        "user": null,
      },
      {
        "_id": "66532802b5e3e6878847239a",
        "comment": "Ab tou chal ja",
        "createdAt": "2024-05-26T12:16:02.423Z",
        "user": null,
      },
      {
        "_id": "665328fd45df2da31e41fc41",
        "comment": "Whyyyyy",
        "createdAt": "2024-05-26T12:20:13.455Z",
        "user": null,
      },
      {
        "_id": "66559e96d0875be818ecf6b2",
        "comment": "Chala?",
        "createdAt": "2024-05-28T09:06:30.745Z",
        "user": null,
      },
      {
        "_id": "66559fb1d0875be818ecf706",
        "comment": "Barelvi",
        "createdAt": "2024-05-28T09:11:13.833Z",
        "user": null,
      },
      {
        "_id": "6655a13cd0875be818ecf744",
        "comment": "Chacha",
        "createdAt": "2024-05-28T09:17:48.653Z",
        "user": null,
      },
      {
        "_id": "6655a188d0875be818ecf785",
        "comment": "Chache",
        "createdAt": "2024-05-28T09:19:04.419Z",
        "user": null,
      },
      {
        "_id": "6655a231d0875be818ecf7c8",
        "comment": "..",
        "createdAt": "2024-05-28T09:21:53.294Z",
        "user": null,
      },
      {
        "_id": "6655a33ad0875be818ecf80f",
        "comment": "Bb",
        "createdAt": "2024-05-28T09:26:18.216Z",
        "user": null,
      },
      {
        "_id": "6655a387d0875be818ecf83d",
        "comment": "Gg",
        "createdAt": "2024-05-28T09:27:35.559Z",
        "user": null,
      },
      {
        "_id": "6655a400d0875be818ecf86d",
        "comment": "Hn be?",
        "createdAt": "2024-05-28T09:29:36.497Z",
        "user": null,
      },
      {
        "_id": "6655a50dd2f8480bc8145ed8",
        "comment": "Mum",
        "createdAt": "2024-05-28T09:34:05.863Z",
        "user": null,
      },
      {
        "_id": "6655a543d2f8480bc8145f0c",
        "comment": "Mmm",
        "createdAt": "2024-05-28T09:34:59.827Z",
        "user": null,
      },
      {
        "_id": "6655a9f94ace1fa3db63c654",
        "comment": "How",
        "createdAt": "2024-05-28T09:55:05.311Z",
        "user": null,
      },
      {
        "_id": "6655aa104ace1fa3db63c68d",
        "comment": "Mmm",
        "createdAt": "2024-05-28T09:55:28.954Z",
        "user": null,
      },
      {
        "_id": "66642ac1c008cbfeaf8ab15d",
        "comment": "Loot maar",
        "createdAt": "2024-06-08T09:56:17.750Z",
        "user": null,
      },
    ],
    "createdAt": "2024-04-29T15:28:15.630Z",
    "image": "https://firebasestorage.googleapis.com/v0/b/eco-track-37da4.appspot.com/o/410816f6-823a-4254-bec9-e619dead977c.jpeg?alt=media&token=ba05d0c2-eb01-447e-8708-533b9a32a5b5",
    "isTypeAd": false,
    "likes": [
      "662cad97e4e3ee6c22e99507",
      "66522b44c655d7e2fba0c7be",
    ],
    "postDescription": "",
    "shares": [],
    "updatedAt": "2024-06-12T12:44:00.343Z",
    "user": null,
  },
  {
    "__v": 17,
    "_id": "662f826cbb7fe0ab56da0521",
    "comments": [
      {
        "_id": "665329ed45df2da31e41fc6a",
        "comment": "Hi",
        "createdAt": "2024-05-26T12:24:13.762Z",
        "user": null,
      },
      {
        "_id": "66532b5ab5e3e687884723c7",
        "comment": "Wheo",
        "createdAt": "2024-05-26T12:30:18.113Z",
        "user": null,
      },
      {
        "_id": "66532dd93f22d7aacc0f34d9",
        "comment": "Whoaa",
        "createdAt": "2024-05-26T12:40:57.486Z",
        "user": null,
      },
      {
        "_id": "6655a41bd0875be818ecf88e",
        "comment": "Chacha mumthal kumar",
        "createdAt": "2024-05-28T09:30:03.945Z",
        "user": null,
      },
      {
        "_id": "6655a430d0875be818ecf8a0",
        "comment": "Aziz arif",
        "createdAt": "2024-05-28T09:30:24.389Z",
        "user": null,
      },
      {
        "_id": "6655a49bd0875be818ecf8d3",
        "comment": "Maila",
        "createdAt": "2024-05-28T09:32:11.768Z",
        "user": null,
      },
      {
        "_id": "6655a4cfd2f8480bc8145ea4",
        "comment": "Malir",
        "createdAt": "2024-05-28T09:33:03.608Z",
        "user": null,
      },
      {
        "_id": "6655a4d0d2f8480bc8145ea9",
        "comment": "Malir",
        "createdAt": "2024-05-28T09:33:04.020Z",
        "user": null,
      },
      {
        "_id": "6655a557d2f8480bc8145f34",
        "comment": "Murtaza korangi ",
        "createdAt": "2024-05-28T09:35:19.488Z",
        "user": null,
      },
      {
        "_id": "666ddd4c5c0ca4ea4bf3ad6d",
        "comment": "Hhhh",
        "createdAt": "2024-06-15T18:28:28.333Z",
        "user": null,
      },
    ],
    "createdAt": "2024-04-29T11:20:12.665Z",
    "image": "https://firebasestorage.googleapis.com/v0/b/ecotrack-decb2.appspot.com/o/f63d11f6-9b65-40ae-8039-1e8895fdac77.jpeg?alt=media&token=202ffef1-37c4-403b-85de-10e6145ed2d5",
    "isTypeAd": false,
    "likes": [
      "662cad97e4e3ee6c22e99507",
      "66522b44c655d7e2fba0c7be",
      "65f6cf15a0a6a67d2493d722",
    ],
    "shares": [],
    "updatedAt": "2024-06-19T13:10:07.569Z",
    "user": null,
  },
]
`;

exports[`Post API tests should get all posts of a specific user 1`] = `
[
  {
    "__v": 33,
    "_id": "662fbc8ff9e8c5468deeca7f",
    "comments": [
      {
        "_id": "664450f4501eb2dc65217435",
        "comment": "Yes ",
        "createdAt": "2024-05-15T06:06:44.454Z",
        "user": null,
      },
      {
        "_id": "6652fb3ebe8826f705f240e9",
        "comment": "Nice",
        "createdAt": "2024-05-26T09:05:02.809Z",
        "user": null,
      },
      {
        "_id": "665322deb5e3e687884722c3",
        "comment": "Kya post hai bhai",
        "createdAt": "2024-05-26T11:54:06.751Z",
        "user": null,
      },
      {
        "_id": "6653237a45df2da31e41fbba",
        "comment": "Good",
        "createdAt": "2024-05-26T11:56:42.007Z",
        "user": null,
      },
      {
        "_id": "6653245645df2da31e41fbd6",
        "comment": "Nice",
        "createdAt": "2024-05-26T12:00:22.557Z",
        "user": null,
      },
      {
        "_id": "665325b3b5e3e687884722ea",
        "comment": "Hi",
        "createdAt": "2024-05-26T12:06:11.239Z",
        "user": null,
      },
      {
        "_id": "66532640b5e3e6878847230c",
        "comment": "Try",
        "createdAt": "2024-05-26T12:08:32.296Z",
        "user": null,
      },
      {
        "_id": "665326e245df2da31e41fc12",
        "comment": "Wheo",
        "createdAt": "2024-05-26T12:11:14.032Z",
        "user": null,
      },
      {
        "_id": "6653273db5e3e6878847233e",
        "comment": "Thoko thoko",
        "createdAt": "2024-05-26T12:12:45.668Z",
        "user": null,
      },
      {
        "_id": "665327bdb5e3e6878847236b",
        "comment": "Fuck",
        "createdAt": "2024-05-26T12:14:53.975Z",
        "user": null,
      },
      {
        "_id": "66532802b5e3e6878847239a",
        "comment": "Ab tou chal ja",
        "createdAt": "2024-05-26T12:16:02.423Z",
        "user": null,
      },
      {
        "_id": "665328fd45df2da31e41fc41",
        "comment": "Whyyyyy",
        "createdAt": "2024-05-26T12:20:13.455Z",
        "user": null,
      },
      {
        "_id": "66559e96d0875be818ecf6b2",
        "comment": "Chala?",
        "createdAt": "2024-05-28T09:06:30.745Z",
        "user": null,
      },
      {
        "_id": "66559fb1d0875be818ecf706",
        "comment": "Barelvi",
        "createdAt": "2024-05-28T09:11:13.833Z",
        "user": null,
      },
      {
        "_id": "6655a13cd0875be818ecf744",
        "comment": "Chacha",
        "createdAt": "2024-05-28T09:17:48.653Z",
        "user": null,
      },
      {
        "_id": "6655a188d0875be818ecf785",
        "comment": "Chache",
        "createdAt": "2024-05-28T09:19:04.419Z",
        "user": null,
      },
      {
        "_id": "6655a231d0875be818ecf7c8",
        "comment": "..",
        "createdAt": "2024-05-28T09:21:53.294Z",
        "user": null,
      },
      {
        "_id": "6655a33ad0875be818ecf80f",
        "comment": "Bb",
        "createdAt": "2024-05-28T09:26:18.216Z",
        "user": null,
      },
      {
        "_id": "6655a387d0875be818ecf83d",
        "comment": "Gg",
        "createdAt": "2024-05-28T09:27:35.559Z",
        "user": null,
      },
      {
        "_id": "6655a400d0875be818ecf86d",
        "comment": "Hn be?",
        "createdAt": "2024-05-28T09:29:36.497Z",
        "user": null,
      },
      {
        "_id": "6655a50dd2f8480bc8145ed8",
        "comment": "Mum",
        "createdAt": "2024-05-28T09:34:05.863Z",
        "user": null,
      },
      {
        "_id": "6655a543d2f8480bc8145f0c",
        "comment": "Mmm",
        "createdAt": "2024-05-28T09:34:59.827Z",
        "user": null,
      },
      {
        "_id": "6655a9f94ace1fa3db63c654",
        "comment": "How",
        "createdAt": "2024-05-28T09:55:05.311Z",
        "user": null,
      },
      {
        "_id": "6655aa104ace1fa3db63c68d",
        "comment": "Mmm",
        "createdAt": "2024-05-28T09:55:28.954Z",
        "user": null,
      },
      {
        "_id": "66642ac1c008cbfeaf8ab15d",
        "comment": "Loot maar",
        "createdAt": "2024-06-08T09:56:17.750Z",
        "user": null,
      },
    ],
    "createdAt": "2024-04-29T15:28:15.630Z",
    "image": "https://firebasestorage.googleapis.com/v0/b/eco-track-37da4.appspot.com/o/410816f6-823a-4254-bec9-e619dead977c.jpeg?alt=media&token=ba05d0c2-eb01-447e-8708-533b9a32a5b5",
    "isTypeAd": false,
    "likes": [
      "662cad97e4e3ee6c22e99507",
      "66522b44c655d7e2fba0c7be",
    ],
    "postDescription": "",
    "shares": [],
    "updatedAt": "2024-06-12T12:44:00.343Z",
    "user": null,
  },
  {
    "__v": 4,
    "_id": "6655ab23e25bcfa530d9c27c",
    "comments": [
      {
        "_id": "665923dd60f0ce5aa4e62313",
        "comment": "Good jerry",
        "createdAt": "2024-05-31T01:11:57.899Z",
        "user": null,
      },
    ],
    "createdAt": "2024-05-28T10:00:03.329Z",
    "image": "https://firebasestorage.googleapis.com/v0/b/eco-track-37da4.appspot.com/o/24bc5260-81b3-4fb3-b754-72dd3e135149.jpeg?alt=media&token=03573cf5-847b-4426-9019-2964f7fa2fb5",
    "isTypeAd": false,
    "likes": [
      "65f6cf15a0a6a67d2493d722",
    ],
    "postDescription": "Hi",
    "shares": [],
    "updatedAt": "2024-06-08T09:55:56.800Z",
    "user": null,
  },
  {
    "__v": 2,
    "_id": "666944a2d8761d7fde2a5e5e",
    "comments": [
      {
        "_id": "666d8b56d337baa088f190a3",
        "comment": "Great initiative ",
        "createdAt": "2024-06-15T12:38:46.628Z",
        "user": null,
      },
    ],
    "createdAt": "2024-06-12T06:48:02.251Z",
    "image": "https://firebasestorage.googleapis.com/v0/b/eco-track-37da4.appspot.com/o/adb38d5b-67db-49d8-a08f-266567a60542.jpeg?alt=media&token=8aea0c44-b50f-42ce-a53f-9fe5b5cbdb4f",
    "isTypeAd": false,
    "likes": [
      "65f6cf15a0a6a67d2493d722",
    ],
    "postDescription": "Hey guys help Us to make our city green against 
",
    "shares": [],
    "updatedAt": "2024-06-15T12:38:46.629Z",
    "user": null,
  },
]
`;

exports[`Post API tests should get specific post by ID 1`] = `
{
  "message": "Post not found",
}
`;

exports[`Post API tests should like a post 1`] = `
{
  "message": "User not found",
}
`;
